### PR TITLE
✨ Set up golang context when initiating controller manager

### PR DESCRIPTION
**Pull request by zhanggbj**
**What this PR does / why we need it**:
Set up golang context when initiating controller manager and also pass it down to VSphereCluster controller.
For other controllers, will update in separate PR.
This is part of #2295.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #2295

### CODE REVIEW

File: controllers/vspherecluster_reconciler.go [hunk 1]
```diff
@@ -440,7 +440,7 @@ func (r clusterReconciler) reconcileVSphereClusterWhenAPIServerIsOnline(clusterC
 	go func() {
 		// Block until the target API server is online.
 		clusterCtx.Logger.Info("start polling API server for online check")
-		wait.PollUntilContextCancel(context.Background(), time.Second*1, true, func(context.Context) (bool, error) { return r.isAPIServerOnline(clusterCtx), nil }) //nolint:errcheck
+		wait.PollUntilContextCancel(ctx, time.Second*1, true, func(context.Context) (bool, error) { return r.isAPIServerOnline(ctx, clusterCtx), nil }) //nolint:errcheck
```
Comment (sbueringer) [inline controllers/vspherecluster_reconciler.go:443]: Let's keep background here and in l.455

The problem is that this go routine continues to run after the reconcile is done. So we don't want to cancel them

Comment (zhanggbj) [reply controllers/vspherecluster_reconciler.go:443]: Thanks for catching it!

File: controllers/vspherecluster_reconciler.go [hunk 2]
```diff
@@ -440,7 +440,7 @@ func (r clusterReconciler) reconcileVSphereClusterWhenAPIServerIsOnline(clusterC
 	go func() {
 		// Block until the target API server is online.
 		clusterCtx.Logger.Info("start polling API server for online check")
-		wait.PollUntilContextCancel(context.Background(), time.Second*1, true, func(context.Context) (bool, error) { return r.isAPIServerOnline(clusterCtx), nil }) //nolint:errcheck
+		wait.PollUntilContextCancel(context.Background(), time.Second*1, true, func(context.Context) (bool, error) { return r.isAPIServerOnline(ctx, clusterCtx), nil }) //nolint:errcheck
```
Comment (sbueringer) [inline controllers/vspherecluster_reconciler.go:443]: We also have to use a background context for isAPIServerOnline and the other func below. Let's go with:

```go
		ctx := context.Background()
		// Block until the target API server is online.
		clusterCtx.Logger.Info("start polling API server for online check")
		wait.PollUntilContextCancel(ctx, time.Second*1, true, func(context.Context) (bool, error) { return r.isAPIServerOnline(ctx, clusterCtx), nil }) //nolint:errcheck
		clusterCtx.Logger.Info("stop polling API server for online check")
		clusterCtx.Logger.Info("triggering GenericEvent", "reason", "api-server-online")
		eventChannel := clusterCtx.GetGenericEventChannelFor(clusterCtx.VSphereCluster.GetObjectKind().GroupVersionKind())
		eventChannel <- event.GenericEvent{
			Object: clusterCtx.VSphereCluster,
		}

		// Once the control plane has been marked as initialized it is safe to
		// remove the key from the map that prevents multiple goroutines from
		// polling the API server to see if it is online.
		clusterCtx.Logger.Info("start polling for control plane initialized")
		wait.PollUntilContextCancel(ctx, time.Second*1, true, func(context.Context) (bool, error) { return r.isControlPlaneInitialized(ctx, clusterCtx), nil }) //nolint:errcheck
		clusterCtx.Logger.Info("stop polling for control plane initialized")
		apiServerTriggersMu.Lock()
		delete(apiServerTriggers, clusterCtx.Cluster.UID)
		apiServerTriggersMu.Unlock()
```

File: main.go [hunk 3]
```diff
@@ -303,17 +303,19 @@ func main() {
...
 		setupLog.Error(err, "problem creating controller manager")
 		os.Exit(1)
 	}
 
 	setupChecks(mgr)
 
-	sigHandler := ctrlsig.SetupSignalHandler()
-	setupLog.Info("starting controller manager")
-	if err := mgr.Start(sigHandler); err != nil {
+	if err := mgr.Start(ctx); err != nil {
```
Comment (sbueringer) [inline main.go:318]: let's keep the log line

File: test/helpers/envtest.go [hunk 4]
```diff
@@ -165,7 +165,7 @@ func NewTestEnvironment() *TestEnvironment {
 		return (&webhooks.VSphereFailureDomainWebhook{}).SetupWebhookWithManager(mgr)
 	}
 
-	mgr, err := manager.New(managerOpts)
+	mgr, err := manager.New(context.Background(), managerOpts)
```
Comment (sbueringer) [inline test/helpers/envtest.go:168]:
```suggestion
	mgr, err := manager.New(ctrl.SetupSignalHandler(), managerOpts)
```
It's a bit hard to tell but in CAPI we are also using this one. Should lead to a better cleanup

**Review by sbueringer — COMMENTED** [review]
Just a few nits

File: main.go (hunk 3 above)
Comment (zhanggbj) [reply main.go:318]: Fixed!

File: controllers/vspherecluster_reconciler.go (hunk 2 above)
Comment (zhanggbj) [reply controllers/vspherecluster_reconciler.go:443]: Thanks for pointing it out!

File: test/helpers/envtest.go (hunk 4 above)
Comment (zhanggbj) [reply test/helpers/envtest.go]: Sure, let's go with it.

Comment (zhanggbj) [reply test/helpers/envtest.go]: With this change, need to update NewTestEnvironment() to consume an existed ctx (initialized in controllers/controller_suite_test.go using ctrl.SetupSingalHandler()), as we cannot initialized ctrl.SetupSingalHandler() for multiple times.

Comment (sbueringer) [reply test/helpers/envtest.go]: Sounds good, thx!

**Comment by sbueringer** [comment]
Thank you very much!!

/lgtm

/assign @chrischdi 
for a final review

**Comment by chrischdi** [comment]
Awesome work! Thanks @zhanggbj 🏆

**Pull request merged**